# March Machine Learning Mania 2017

https://www.kaggle.com/c/march-machine-learning-mania-2017/overview

### What to predict

Stage 1 - You should submit predicted probabilities for every possible matchup in the past 4 NCAA tournaments (2013-2016).

Stage 2 - You should submit predicted probabilities for every possible matchup before the 2017 tournament begins.

Refer to the Timeline page for specific dates. In both stages, the sample submission will tell you which games to predict.

### About March madness

 - Tournament is made of 64 teams competing in Single Elimination Format.
 - To win the championship, team has to win 6 consecutive games. 
 - Tournament is broken into 4 regions. Each region has 16 teams, ranked from 1 - 16. This is determined by NCAA committee
 - Games are structured so that HIghest Seed in a region plays Lowest seed, 2<sup>nd</sup> highest plays with 2<sup>nd</sup> lowest and so on.
 - 

## Data Wrangling

In [2]:
import numpy as np
import pandas as pd

In [3]:
seasons_path = './data/seasons.csv'
teams_path = './data/teams.csv'
reg_season_compact_results_path = './data/RegularSeasonCompactResults.csv'
reg_season_detailed_results_path = './data/RegularSeasonDetailedResults.csv'
tourney_compact_results_path = './data/TourneyCompactResults.csv'
tourney_detailed_results_path = './data/TourneyDetailedResults.csv'
tourney_seeds_path = './data/TourneySeeds.csv'
tourney_slots_path = './data/TourneySlots.csv'

In [4]:
seasons_df = pd.read_csv(seasons_path)
teams_df = pd.read_csv(teams_path)
RSCR_df = pd.read_csv(reg_season_compact_results_path)
RSDR_df = pd.read_csv(reg_season_detailed_results_path)
tcr_df = pd.read_csv(tourney_compact_results_path)
tdr_df = pd.read_csv(tourney_detailed_results_path)
tseeds_df = pd.read_csv(tourney_seeds_path)
tslots_df = pd.read_csv(tourney_slots_path)

### Data Descriptions

***Seasons***

This file identifies the different seasons included in the historical data, along with certain season-level properties.

 - "season" - indicates the year in which the tournament was played
 - "dayzero" - tells you the date corresponding to daynum=0 during that season. All game dates have been aligned upon a common scale so that the championship game of the final tournament is on daynum=154. Working backward, the national semifinals are always on daynum=152, the "play-in" games are on days 134/135, Selection Sunday is on day 132, and so on. All game data includes the day number in order to make it easier to perform date calculations. If you really want to know the exact date a game was played on, you can combine the game's "daynum" with the season's "dayzero". For instance, since day zero during the 2011-2012 season was 10/31/2011, if we know that the earliest regular season games that year were played on daynum=7, they were therefore played on 11/07/2011.
 - "regionW/X/Y/Z" - by convention, the four regions in the final tournament are always named W, X, Y, and Z. Whichever region's name comes first alphabetically, that region will be Region W. And whichever Region plays against Region W in the national semifinals, that will be Region X. For the other two regions, whichever region's name comes first alphabetically, that region will be Region Y, and the other will be Region Z. This allows us to identify the regions and brackets in a standardized way in other files. For instance, during the 2012 tournament, the four regions were East, Midwest, South, and West. Being the first alphabetically, East becomes W. Since the East regional champion (Ohio State) played against the Midwest regional champion (Kansas) in the national semifinals, that makes Midwest be region X. For the other two (South and West), since South comes first alphabetically, that makes South Y and therefore West is Z. So for this season, the W/X/Y/Z are East,Midwest,South,West.

In [5]:
seasons_df.head()

Season     Dayzero Regionw    Regionx    Regiony    Regionz
0    1985  10/29/1984    East       West    Midwest  Southeast
1    1986  10/28/1985    East    Midwest  Southeast       West
2    1987  10/27/1986    East  Southeast    Midwest       West
3    1988   11/2/1987    East    Midwest  Southeast       West
4    1989  10/31/1988    East       West    Midwest  Southeast

***Teams***

This file identifies the different college teams present in the dataset. Each team has a 4 digit id number.

In [7]:
#teams_df.head()
len(teams_df['Team_Name'])

364

***RegularSeasonCompactResults***

This file identifies the game-by-game results for 32 seasons of historical data, from 1985 to 2015. Each year, it includes all games played from daynum 0 through 132 (which by definition is "Selection Sunday," the day that tournament pairings are announced). Each row in the file represents a single game played.

 - "season" - this is the year of the associated entry in seasons.csv (the year in which the final tournament occurs)
 - "daynum" - this integer always ranges from 0 to 132, and tells you what day the game was played on. It represents an offset from the "dayzero" date in the "seasons.csv" file. For example, the first game in the file was daynum=20. Combined with the fact from the "season.csv" file that day zero was 10/29/1984, that means the first game was played 20 days later, or 11/18/1984. There are no teams that ever played more than one game on a given date, so you can use this fact if you need a unique key. In order to accomplish this uniqueness, we had to adjust one game's date. In March 2008, the SEC postseason tournament had to reschedule one game (Georgia-Kentucky) to a subsequent day, so Georgia had to actually play two games on the same day. In order to enforce this uniqueness, we moved the game date for the Georgia-Kentucky game back to its original date.
 - "wteam" - this identifies the id number of the team that won the game, as listed in the "teams.csv" file. No matter whether the game was won by the home team or visiting team, "wteam" always identifies the winning team.
 - "wscore" - this identifies the number of points scored by the winning team.
 - "lteam" - this identifies the id number of the team that lost the game.
 - "lscore" - this identifies the number of points scored by the losing team.
 - "numot" - this indicates the number of overtime periods in the game, an integer 0 or higher.
 - "wloc" - this identifies the "location" of the winning team. If the winning team was the home team, this value will be "H". If the winning team was the visiting team, this value will be "A". If it was played on a neutral court, then this value will be "N". Sometimes it is unclear whether the site should be considered neutral, since it is near one team's home court, or even on their court during a tournament, but for this determination we have simply used the Kenneth Massey data in its current state, where the "@" sign is either listed with the winning team, the losing team, or neither team.

In [6]:
RSCR_df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

***RegularSeasonDetailedResults***

This file is a more detailed set of game results, covering seasons 2003-2016. This includes team-level total statistics for each game (total field goals attempted, offensive rebounds, etc.) The column names should be self-explanatory to basketball fans (as above, "w" or "l" refers to the winning or losing team):

 - wfgm - field goals made
 - wfga - field goals attempted
 - wfgm3 - three pointers made
 - wfga3 - three pointers attempted
 - wftm - free throws made
 - wfta - free throws attempted
 - wor - offensive rebounds
 - wdr - defensive rebounds
 - wast - assists
 - wto - turnovers
 - wstl - steals
 - wblk - blocks
 - wpf - personal fouls

In [7]:
RSDR_df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  Wfgm  Wfga  ...  \
0    2003      10   1104      68   1328      62    N      0    27    58  ...   
1    2003      10   1272      70   1393      63    N      0    26    62  ...   
2    2003      11   1266      73   1437      61    N      0    24    58  ...   
3    2003      11   1296      56   1457      50    N      0    18    38  ...   
4    2003      11   1400      77   1208      71    N      0    30    61  ...   

   Lfga3  Lftm  Lfta  Lor  Ldr  Last  Lto  Lstl  Lblk  Lpf  
0     10    16    22   10   22     8   18     9     2   20  
1     24     9    20   20   25     7   12     8     6   16  
2     26    14    23   31   22     9   12     2     5   23  
3     22     8    15   17   20     9   19     4     3   23  
4     16    17    27   21   15    12   10     7     1   14  

[5 rows x 34 columns]

***TourneyCompactResults***

This file identifies the game-by-game NCAA tournament results for all seasons of historical data. The data is formatted exactly like the regular_season_compact_results.csv data. Note that these games also include the play-in games (which always occurred on day 134/135) for those years that had play-in games.

In [8]:
tcr_df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985     136   1116      63   1234      54    N      0
1    1985     136   1120      59   1345      58    N      0
2    1985     136   1207      68   1250      43    N      0
3    1985     136   1229      58   1425      55    N      0
4    1985     136   1242      49   1325      38    N      0

***TourneyDetailedResults***

This file contains the more detailed results for tournament games from 2003 onward.

In [9]:
tdr_df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  Wfgm  Wfga  ...  \
0    2003     134   1421      92   1411      84    N      1    32    69  ...   
1    2003     136   1112      80   1436      51    N      0    31    66  ...   
2    2003     136   1113      84   1272      71    N      0    31    59  ...   
3    2003     136   1141      79   1166      73    N      0    29    53  ...   
4    2003     136   1143      76   1301      74    N      1    27    64  ...   

   Lfga3  Lftm  Lfta  Lor  Ldr  Last  Lto  Lstl  Lblk  Lpf  
0     31    14    31   17   28    16   15     5     0   22  
1     16     7     7    8   26    12   17    10     3   15  
2     28    14    21   20   22    11   12     2     5   18  
3     17    12    17   14   17    20   21     6     6   21  
4     21    15    20   10   26    16   14     5     8   19  

[5 rows x 34 columns]

***TourneySeeds***

This file identifies the seeds for all teams in each NCAA tournament, for all seasons of historical data. Thus, there are between 64-68 rows for each year, depending on the bracket structure.

 - "season" - the year
 - "seed" - this is a 3/4-character identifier of the seed, where the first character is either W, X, Y, or Z (identifying the region the team was in) and the next two digits (either 01, 02, ..., 15, or 16) tells you the seed within the region. For play-in teams, there is a fourth character (a or b) to further distinguish the seeds, since teams that face each other in the play-in games will have the same first three characters. For example, the first record in the file is seed W01, which means we are looking at the #1 seed in the W region (which we can see from the "seasons.csv" file was the East region). This seed is also referenced in the "tourney_slots.csv" file that tells us which bracket slots face which other bracket slots in which rounds.
 - "team" - this identifies the id number of the team, as specified in the teams.csv file.

In [10]:
tseeds_df.head()

Season Seed  Team
0    1985  W01  1207
1    1985  W02  1210
2    1985  W03  1228
3    1985  W04  1260
4    1985  W05  1374

***TourneySlots***

This file identifies the mechanism by which teams are paired against each other, depending upon their seeds. Because of the existence of play-in games for particular seed numbers, the pairings have small differences from year to year. If there were N teams in the tournament during a particular year, there were N-1 teams eliminated (leaving one champion) and therefore N-1 games played, as well as N-1 slots in the tournament bracket, and thus there will be N-1 records in this file for that season.

 - "season" - the year
 - "slot" - this uniquely identifies one of the tournament games. For play-in games, it is a three-character string identifying the seed fulfilled by the winning team, such as W16 or Z13. For regular tournament games, it is a four-character string, where the first two characters tell you which round the game is (R1, R2, R3, R4, R5, or R6) and the second two characters tell you the expected seed of the favored team. Thus the first row is R1W1, identifying the Round 1 game played in the W bracket, where the favored team is the 1 seed. As a further example, the R2W1 slot indicates the Round 2 game that would have the 1 seed from the W bracket, assuming that all favored teams have won up to that point. The slot names are different for the final two rounds, where R5WX identifies the national semifinal game between the winners of regions W and X, and R5YZ identifies the national semifinal game between the winners of regions Y and Z, and R6CH identifies the championship game. The "slot" value is used in other columns in order to represent the advancement and pairings of winners of previous games.
 - "strongseed" - this indicates the expected stronger-seeded team that plays in this game. For Round 1 games, a team seed is identified in this column (as listed in the "seed" column in the tourney_seeds.csv file), whereas for subsequent games, a slot is identified in this column. In the first record of this file (slot R1W1), we see that seed W01 is the "strongseed", which during the 1985 tournament would have been Georgetown. Whereas for games from Round 2 or later, rather than a team seed, we will see a "slot" referenced in this column. So in the 33rd record of this file (slot R2W1), it tells us that the winners of slots R1W1 and R1W8 will face each other in Round 2. Of course, in the last few games of the tournament - the national semifinals and finals - it's not really meaningful to talk about a "strong seed" or "weak seed", but those games are represented in the same format for the sake of uniformity.
 - "weakseed" - this indicates the expected weaker-seeded team that plays in this game, assuming all favored teams have won so far. For Round 1 games, a team seed is identified in this column (as listed in the "seed" column in the tourney_seeds.csv file), whereas for subsequent games, a slot is identified in this column.

In [11]:
tslots_df.head()

Season  Slot Strongseed Weakseed
0    1985  R1W1        W01      W16
1    1985  R1W2        W02      W15
2    1985  R1W3        W03      W14
3    1985  R1W4        W04      W13
4    1985  R1W5        W05      W12

In [12]:
teams_df.shape

(364, 2)

In [13]:
seasons_df.shape

(33, 6)

### Inferences from Data Wrangling

Teams : 364 teams

seasons : 33 seasons (1985 - 2017)

Team_Id      Team_Name
0       1101    Abilene Chr
1       1102      Air Force
2       1103          Akron
3       1104        Alabama
4       1105    Alabama A&M
..       ...            ...
359     1460      Wright St
360     1461        Wyoming
361     1462         Xavier
362     1463           Yale
363     1464  Youngstown St

[364 rows x 2 columns]